### PR TITLE
Bugfix/get secret logged out

### PR DESCRIPTION
**Pull request by itfranck**
It looks like I inverted the IgnoreError part in the main branch when I did the rework on Invoke-Lpass.
It caused the prompt not to show (again)

While working on a solution, I entered other issues with the stream redirection.
During debug, it seemed that calling Invoke-Lpass with the redirect reacted differently than applying it on the lpass CLI result itself.

Lpass cli result returned $null in some occasion where doing it on Invoke-Lpass worked consistently no matter what the call.

Therefore, I created a function where I wrapped the call logic and used that within Invoke-lpass.

Not only the issue is fixed but the whole thing is improved. 

`-ErrorAction SilentlyContinue` is not used anymore.
I found that a bit confusing since in all cases, we generated errors (CLI or otherwise) from the function. 
Instead, this is managed depending on the command with more granularity. 

- login
Always show prompt and thus always show native error (if any)

- show
**Before**
- command not redirected (for prompt if needed)
- Native error shown in console
- Secret was not found error shown even though it should be a logged out error (since we didn't capture the CLI output and left it through) 

**Now**
- Command is redirected
- We check for the logged out message
- If we get it, we do a second call without redirection and we'll get one of the following results:
1: Prompt (because the session was disconnected)
2: native error + logged out message (because we are truly logged out) 

If you are connected, you won't get the native error showing anymore either.

### CODE REVIEW

File: SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1 [hunk 1]
```diff
@@ -52,16 +52,53 @@ $SpecialKeys = @('Language', 'NoteType', 'Notes')
...
-        $Arguments,
-
+        [string[]]$Arguments,
         [Parameter(ValueFromPipeline)]
-        [object]
-        $InputObject,
+        [object]$InputObject,
         #Only for root module functions
         [hashtable]$VaultParams
     )
```
Comment (itfranck) [inline SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:104]: Half of the parameter were on the same line `[String]$Param` and the other was dual line. 
So I made that consistent. I personally prefer having the type and param on the same line (easier to read / less scrolling) but if you prefer the other way around, I can change it so it's the other way around (but consistent)

File: SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1 [hunk 2]
```diff
@@ -70,35 +107,51 @@ function Invoke-lpass {
...
-            {$UseNative}                     {$result = $InputObject | & $lpassPath @Arguments; break}
-            {$UseWSL -and $IgnoreErrors}     {$result = $InputObject | & wsl $lpassPath @Arguments 2>&1; break}
-            {$UseWSL}                        {$result = $InputObject | & wsl $lpassPath @Arguments; break}
-        }
-    } else {
-        switch ($true) {
-            { $UseNative -and $IgnoreErrors }   {$result = & $lpassPath @Arguments 2>&1; break}
-            { $UseNative }                      {$result = & $lpassPath @Arguments; break}
-            { $UseWSL -and $IgnoreErrors }      {$result = & wsl $lpassPath @Arguments 2>&1; break}
-            { $UseWSL }                         {$result = & wsl $lpassPath @Arguments; break}
```
Comment (itfranck) [inline SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:95]: Moved to Invoke-lpassInternal for reason higlighted in PR

Comment (itfranck) [reply SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:95]: Also, it reduce the overall duplication of code...

File: SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1 [hunk 3]
```diff
@@ -70,35 +107,51 @@ function Invoke-lpass {
...
+                    # If $result2 -eq $null, something will have been printed in the console (because we disabled the redirect)
+                    # We therefore want to keep the original $result "Logged out" so it is thrown later on
+                    # If not $null, we want to evaluate $result2 instead and discard $result
+                    if ($null -ne $result2) {$result = $result2}
+                }
+            }
+            Default {
+                # By default, we always redirect streams
+                $result = Invoke-lpassInternal @Params 2>&1
+            }
```
Comment (itfranck) [inline SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:148]: That's the bit  I was talking about.
By default, redirect all.
Login = No redirect ever (used by Connect-LastPass)
Show = Redirect first, check for logged out message and only if that occurs, redo the command without redirect so prompt is allowed.

Comment (itfranck) [reply SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:148]: We could even have gone further and do a login if we got $null, meaning the user was likely logged out and it would have mostly worked perfectly except in the scenario where a disconnected user cancelled the password prompt on purpose. In that case, that would do a password prompt then ask you for your user and a second password prompt.

File: SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1 [hunk 4]
```diff
@@ -70,35 +107,51 @@ function Invoke-lpass {
...
         }
     }
 
     if ($result -is [System.Management.Automation.ErrorRecord]) {
         switch -Wildcard ([string] $result) {
-            $lpassMessage.LoggedOut { throw [PasswordRequiredException] $lpassMessage.LoggedOut }
+            $lpassMessage.LoggedOut { throw [PasswordRequiredException] "$($lpassMessage.LoggedOut.TrimEnd("*")) Connect-LastPass" }
```
Comment (itfranck) [inline SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:154]: That's what I intended previously when I modified the `$lpassMessage.LoggedOut` string withotu consideration for its use in the switch statement.

File: SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1 [hunk 5]
```diff
@@ -137,7 +190,7 @@ function Get-Secret
         $Name = $Matches[1]
     }
 
-    $res = Invoke-lpass 'show', '--name', $Name, '--all' -ErrorAction SilentlyContinue
+    $res = Invoke-lpass 'show', '--name', $Name, '--all'
```
Comment (itfranck) [inline SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:193]: These are not needed anymore as we deal with everything at the command level.
We only want to disable redirection for login (always) and show (conditionnally) so this is not needed anymore.

**Comment by itfranck** [comment]
After that, I don't think we'll need to touch `Invoke-lpass` anymore (hopefully)

**Comment by itfranck** [comment]
^ Rebased From Main

**Comment by itfranck** [comment]
@TylerLeonhardt 
If you want to review, this one is done. 
I don't intend to make any more change and couldn't find any remaining issues.

File: SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1 [hunk 6]
```diff
@@ -220,7 +273,7 @@ function Set-Secret
         }
     }
 
-    if ($Secret -is [hashtable]){
+    if ($Secret -is [System.Collections.Specialized.OrderedDictionary] -or $Secret -is [hashtable]) {
```
Comment (TylerLeonhardt) [inline SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1]: Why did you add ordereddictionary here?

File: SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1 [hunk 7]
```diff
@@ -70,35 +107,51 @@ function Invoke-lpass {
...
+                # We want the prompt always, so no redirect
+                $result = Invoke-lpassInternal @Params
+            }
+            {$_ -in 'show','ls','rm'} {
+                # We might want the prompt, but are not sure yet.
+                $result = Invoke-lpassInternal @Params 2>&1
+                # If we get the message stating we might be logged out, we reissue the command without redirect (for Prompt)
+                if ($result -is [System.Management.Automation.ErrorRecord] -and [String]$result -like $lpassMessage.LoggedOut) {
+                    $result2 = Invoke-lpassInternal @Params
+                    # If $result2 -eq $null, something will have been printed in the console (because we disabled the redirect)
```
Comment (TylerLeonhardt) [inline SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:139]: Can we still throw that SecretManagement special exception when they didn't put their password in so that a user could theoretically use a try/catch around Set-Secret?

File: SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1 (hunk 6 above)
Comment (itfranck) [reply SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1]: @TylerLeonhardt 
I forgot to add a comment.
In fact, I forgot about it and had to check in the code.

It is a missing check from : 
https://github.com/TylerLeonhardt/SecretManagement.LastPass/pull/22

Since that PR implemented the ordered dictionary to preserve the order of parameters, everywhere that checked previously for a hashtable now need to check for an ordered dictionary.
(All other instances of this were covered by aforementionned PR)

That being said, I also kept legacy in case someone wanted to build from the ground up a secret using a hashtable and push it to LastPass.

File: SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1 (hunk 7 above)
Comment (itfranck) [reply SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:139]: @TylerLeonhardt If I get your gist, this is what line 142 address. 

```
 if ($null -ne $result2 -or $Arguments[0] -eq 'rm') { $result = $result2 }
```

Basically, we do this : 

1. First call (redirect) is made
2. Check for logged out is made
3. Call is made again without suppression of CLI native error (to allow prompt)
4. Line #142

If the command is rm, success is assumed (This is because rm return $null)
If the command is anything else (show, ls) , then a $null check is made.
- If $result2 is $null, we do nothing so $result (which contains the original logged out redirected error) is the one evaluated later and an error is thrown to this effect.

- If $result2 is not $null, we assign its value to $result so an error is NOT thrown later and everything continue.

As for `Set-Secret` There's always a `show` first to determine if we are in add or edit mode.
Thus, the prompt is evaluated then and since we throw an error, the Set secret never go past the show operation and work properly.

File: SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1 (hunk 6 above)
Comment (TylerLeonhardt) [reply SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1]: So [Ordereddictionary](https://docs.microsoft.com/en-us/dotnet/api/system.collections.specialized.ordereddictionary?view=net-5.0) and [hashtable](https://docs.microsoft.com/en-us/dotnet/api/system.collections.hashtable?view=net-5.0) inherit from [IDictionary](https://docs.microsoft.com/en-us/dotnet/api/system.collections.idictionary?view=net-5.0) so if you replace this with just one `-is [System.Collections.IDictionary]` it should still work.

Comment (TylerLeonhardt) [reply SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1]: This also means that if a user passes in anything that inherits from IDictionary, we will support it :)

File: SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1 (hunk 7 above)
Comment (TylerLeonhardt) [reply SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:139]: Ahhh I see! Ok :) I guess I need more ☕

File: SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1 (hunk 4 above)
Comment (TylerLeonhardt) [reply SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:154]: Gotcha! I find the `trim('*')` kinda ugly though...plus that trim will have to be rerun every single time... I think maybe we can use a new constant variable for and I'll be a tad clean. That also gives us a chance to change the text altogether. I'm not the biggest fan of this part that says `Could not find decryption key`. Maybe say "It appears you're not logged in, try logging in with Connect-LastPass" or something like that.

**Review by TylerLeonhardt — COMMENTED** [review]
A few other tiny things and this is good!

File: SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1 (hunk 6 above)
Comment (itfranck) [reply SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1]: Ohhhhh 👍 Done.

**Comment by itfranck** [comment]
@TylerLeonhardt  Done.

**Comment by TylerLeonhardt** [comment]
Added a small fix for the tests which were throwing an error since $Params was not defined.

**Review by TylerLeonhardt — APPROVED** [review]
LGTM
